src/ucp/core/ucp_worker.c

```diff
@@ -1213,13 +1213,87 @@ static void ucp_worker_close_ifaces(ucp_worker_h worker)
     UCS_ASYNC_UNBLOCK(&worker->async);
 }
 
+static ucs_status_t
+ucp_worker_get_sys_device_distance(ucp_context_h context,
+                                   ucp_rsc_index_t rsc_index,
+                                   ucs_sys_dev_distance_t *distance)
+{
+    ucp_rsc_index_t md_index1, md_index2;
+    uct_tl_resource_desc_t *tl_resources;
+    ucs_sys_device_t device1, device2;
+    unsigned num_tl_resources;
+    unsigned i;
+    ucs_status_t status;
+    ucp_tl_md_t *md;
+
+    device1   = UCS_SYS_DEVICE_ID_UNKNOWN;
+    device2   = UCS_SYS_DEVICE_ID_UNKNOWN;
+    md_index1 = context->tl_rscs[rsc_index].md_index;
+    md        = &context->tl_mds[md_index1];
+
+    status = uct_md_query_tl_resources(md->md, &tl_resources, &num_tl_resources);
+    if (status != UCS_OK) {
+        ucs_error("Failed to query resources: %s", ucs_status_string(status));
+        goto err;
+    }
+
+    for (i = 0; i < num_tl_resources; i++) {
+        if (!strcmp(context->tl_rscs[rsc_index].tl_rsc.dev_name, tl_resources[i].dev_name)) {
+            device1 = tl_resources[i].sys_device;
+            break;
+        }
+    }
+
+    if (device1 == UCS_SYS_DEVICE_ID_UNKNOWN) {
+        goto err_free;
+    }
+
+    uct_release_tl_resource_list(tl_resources);
+
+    for (md_index2 = 0; md_index2 < context->num_mds; ++md_index2) {
+        if (!strcmp(context->tl_mds[md_index2].rsc.md_name,
+                    context->config.ext.selection_cmp)) {
+            md = &context->tl_mds[md_index2];
+
+            status = uct_md_query_tl_resources(md->md, &tl_resources, &num_tl_resources);
+            if (status != UCS_OK) {
+                ucs_error("Failed to query resources: %s", ucs_status_string(status));
+                goto err;
+            }
+
+            if (num_tl_resources < 1) {
+                ucs_error("Unable to find tl_resources for %s comparison",
+                          context->config.ext.selection_cmp);
+                goto err_free;
+            }
+
+            device2 = tl_resources[0].sys_device;
```

Comment:
find the first rsc_index in context->tl_rscs so that context->tl_rscs[rsc_index].md_index == md_index